usersession/userd/launcher.go

```diff
@@ -123,6 +131,105 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntry implements the 'OpenDesktopEntry' method of the 'io.snapcraft.Launcher'
+// DBus interface. Before the provided desktop_file is parsed it is validated against a list
+// of allowed locations.
+func (s *Launcher) OpenDesktopEntry(desktop_file_id string, sender dbus.Sender) *dbus.Error {
+
+	return s.OpenDesktopEntryEnv(desktop_file_id, []string{}, sender)
+}
+
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface.
+func (s *Launcher) OpenDesktopEntryEnv(desktop_file_id string, env []string, sender dbus.Sender) *dbus.Error {
+	exec_command, err := s.readExecCommandFromDesktopFile(s.desktopFileIdToFilename(desktop_file_id))
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	// Passing exec variables between confined snaps raises unanswered questions and they are not required
+	// for the simple cases.  For now, we don't have support for passing them in the dbus API and truncate
+	// the exec command at the first exec variable.
+	// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+	exec_command = strings.SplitN(exec_command, "%", 2)[0]
+
+	args := strings.Split(exec_command, " ")
+	cmd := exec.Command(args[0], args[1:]...)
+
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	if err := cmd.Start(); err != nil {
```

Comment:
It looks like nothing is going to `Wait()` on this command, so seems liable to leave zombie processes behind.

Long term, the spawned process probably shouldn't have userd as the parent.  That could be done through the double-fork technique or perhaps find some way to have systemd spawn it.

Reply:
If we are going this route should userd become a reaper for children? Does this impact the desire to auto-quit from userd on idle?

Reply:
NOTE: @jhenstridge is spot on and we must call Wait or solve this some other way. Perhaps we should ask systemd to run the command?

Reply:
I've not yet had time to look into this, but the code was copy-pasted from the existing `OpenURL()` which presumably has the propensity to create zombies. (Proxying through `xdg-open` avoids the parenting issue.)

Reply:
The `OpenURL` code calls `Cmd.Run()`, which is essentially `Start()` and `Wait()` combined into a single call.

It isn't as simple as switching your code over to `Run()` though, since the processes you're launching are going to be long lived.  And if userd gained "exit on idle" behaviour, perhaps even longer than userd itself.

Reply:
Ack. I've not found a canonical way to fork/exec in Go. I'll have another look tomorrow and if my google foo fails I'll roll my own.

Reply:
I'm not exactly proud of [this solution](https://github.com/snapcore/snapd/pull/7490/commits/a0ec47c6479408e33d25973aefe2e5e9ecf127be) to the parenting issues but it seems to work without fighting the Go libraries. If there's a better way, then please tell me about it.

Reply:
Your solution is effectively the double-fork technique I mentioned above with the shell being used as the immediate child process, which should solve the zombie problem.  It'd be nice to have something that didn't open us to the possibility of shell argument parsing bugs though.